chpl/chpl-service/src/main/java/gov/healthit/chpl/scheduler/job/urlStatus/QuestionableUrlReportGenerator.java

```diff
@@ -5,423 +5,258 @@
-import gov.healthit.chpl.dao.CertificationResultDetailsDAO;
 import gov.healthit.chpl.dao.CertifiedProductDAO;
-import gov.healthit.chpl.dao.DeveloperDAO;
-import gov.healthit.chpl.dao.TestingLabDAO;
-import gov.healthit.chpl.dto.CertificationBodyDTO;
-import gov.healthit.chpl.dto.CertificationResultDetailsDTO;
-import gov.healthit.chpl.dto.CertifiedProductSummaryDTO;
-import gov.healthit.chpl.dto.DeveloperDTO;
-import gov.healthit.chpl.dto.TestingLabDTO;
+import gov.healthit.chpl.dto.CertifiedProductDetailsDTO;
+import gov.healthit.chpl.entity.CertificationStatusType;
 import gov.healthit.chpl.exception.EntityRetrievalException;
+import gov.healthit.chpl.manager.SchedulerManager;
 import gov.healthit.chpl.scheduler.job.QuartzJob;
+import gov.healthit.chpl.scheduler.job.urlStatus.data.UrlCheckerDao;
+import gov.healthit.chpl.scheduler.job.urlStatus.data.UrlResult;
+import gov.healthit.chpl.scheduler.job.urlStatus.data.UrlType;
+import gov.healthit.chpl.scheduler.job.urlStatus.email.FailedUrlCsvFormatter;
+import gov.healthit.chpl.scheduler.job.urlStatus.email.FailedUrlResult;
+import gov.healthit.chpl.scheduler.job.urlStatus.email.QuestionableUrlLookupDao;
 import gov.healthit.chpl.util.EmailBuilder;
+import lombok.extern.log4j.Log4j2;
 
+@Log4j2(topic = "questionableUrlReportGeneratorJobLogger")
 public class QuestionableUrlReportGenerator extends QuartzJob {
-    private static final Logger LOGGER = LogManager.getLogger("questionableUrlReportGeneratorJobLogger");
-    private static final String[] CSV_HEADER = {
-            "URL", "Status Code", "Status Name", "Error Message", "URL Type", "ONC-ATL", "ONC-ACB", "Developer",
-            "Developer Contact Name", "Developer Contact Email", "Developer Contact Phone Number", "Product", "Version",
-            "CHPL Product Number", "Edition", "Certification Date", "Certification Status", "Criteria",
-            "Date Last Checked"
-    };
-
     @Autowired
     private Environment env;
 
     @Autowired
-    private UrlCheckerDao urlCheckerDao;
-
-    @Autowired
-    private TestingLabDAO atlDao;
+    private CertifiedProductDAO cpDao;
 
     @Autowired
     private CertificationBodyDAO acbDao;
 
     @Autowired
-    private DeveloperDAO devDao;
-
-    @Autowired
-    private CertifiedProductDAO cpDao;
+    private UrlCheckerDao urlCheckerDao;
 
     @Autowired
-    private CertificationResultDetailsDAO certResultDao;
+    private QuestionableUrlLookupDao urlLookupDao;
 
-    @Autowired
-    private JpaTransactionManager txnMgr;
+    private FailedUrlCsvFormatter csvFormatter = new FailedUrlCsvFormatter();
 
     @Override
     public void execute(JobExecutionContext jobContext) throws JobExecutionException {
         SpringBeanAutowiringSupport.processInjectionBasedOnCurrentContext(this);
         LOGGER.info("********* Starting the Questionable URL Report Generator job. *********");
 
-        TransactionTemplate txnTemplate = new TransactionTemplate(txnMgr);
```

Comment:
Is this transaction no longer needed? I thought we had to wrap something in these execute methods in a transaction for some reason, but I'm not sure what that was. Maybe only for ones that are saving data?

Reply:
This code was all wrapped in a transaction because it called a bunch of DAO methods in other classes that were not marked Transactional (not, as far as I could tell, because it all has to be completed within a single transaction for data integrity). I refactored all of those calls into a separate DAO - QuestionableUrlLookupDao that is used by this job and does have Transactional annotations.